### PR TITLE
ARROW-8279: [C++] Do not export Codec implementation symbols, remove codec-specific headers

### PR DESCRIPTION
**Pull request by wesm**
This reduces a lot of boilerplate associated with Codec implementation.

### CODE REVIEW

File: cpp/src/arrow/util/compression.h [hunk 1]
```diff
@@ -174,5 +174,55 @@ class ARROW_EXPORT Codec {
   virtual Status Init();
 };
 
+// ----------------------------------------------------------------------
+// Codec factories
+
+namespace internal {
```
Comment (wesm) [inline cpp/src/arrow/util/compression.h]: These could also be moved to "compression_internal.h" if they are seen as a nuisance here

**Comment by wesm** [comment]
The test failures here are ARROW-8280

File: cpp/src/arrow/util/compression.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/compression.h]: That would be nice IMHO.

File: cpp/src/arrow/util/compression_bz2.cc [hunk 2]
```diff
@@ -37,6 +37,7 @@
 
 namespace arrow {
 namespace util {
+namespace internal {
```
Comment (pitrou) [inline cpp/src/arrow/util/compression_bz2.cc]: Can put most of this file in the anonymous namespace actually.

File: cpp/src/arrow/util/compression_lz4.cc [hunk 3]
```diff
@@ -34,6 +34,7 @@
 
 namespace arrow {
 namespace util {
+namespace internal {
```
Comment (pitrou) [inline cpp/src/arrow/util/compression_lz4.cc]: Ditto: anonymous namespace?

**Review by pitrou — COMMENTED** [review]
Well spotted :-) I think all class implementations can be moved to the anonymous namespace actually, so that only factory functions stay exposed.

**Comment by wesm** [comment]
Roger that, I'll move things to anonymous namespaces and add the internal header.

**Comment by wesm** [comment]
+1. Will merge on green

**Comment by wesm** [comment]
Hm grpc_ep continuing to fail in the MinGW builds

```
CMake Error at D:/a/arrow/arrow/build/cpp/grpc_ep-prefix/src/grpc_ep-stamp/grpc_ep-build-RELEASE.cmake:62 (message):
  Command failed: 2

   'D:/a/arrow/msys64/usr/bin/make'

  See also

    D:/a/arrow/arrow/build/cpp/grpc_ep-prefix/src/grpc_ep-stamp/grpc_ep-build-*.log


make[2]: *** [CMakeFiles/grpc_ep.dir/build.make:130: grpc_ep-prefix/src/grpc_ep-stamp/grpc_ep-build] Error 1
make[1]: *** [CMakeFiles/Makefile2:1066: CMakeFiles/grpc_ep.dir/all] Error 2
make: *** [Makefile:158: all] Error 2
```

**Comment by wesm** [comment]
Another flaky googletest download. Hopefully this stops happening

```
FAILED: googletest_ep-prefix/src/googletest_ep-stamp/googletest_ep-download 
cmd.exe /C "cd /D C:\projects\arrow\cpp\build\googletest_ep-prefix\src && C:\Miniconda37-x64\envs\arrow\Library\bin\cmake.exe -P C:/projects/arrow/cpp/build/googletest_ep-prefix/src/googletest_ep-stamp/googletest_ep-download-RELEASE.cmake && C:\Miniconda37-x64\envs\arrow\Library\bin\cmake.exe -E touch C:/projects/arrow/cpp/build/googletest_ep-prefix/src/googletest_ep-stamp/googletest_ep-download"
CMake Error at googletest_ep-stamp/googletest_ep-download-RELEASE.cmake:49 (message):
  Command failed: 1
   'C:/Miniconda37-x64/envs/arrow/Library/bin/cmake.exe' '-Dmake=' '-Dconfig=' '-P' 'C:/projects/arrow/cpp/build/googletest_ep-prefix/src/googletest_ep-stamp/googletest_ep-download-RELEASE-impl.cmake'
  See also
    C:/projects/arrow/cpp/build/googletest_ep-prefix/src/googletest_ep-stamp/googletest_ep-download-*.log
[9/269] Performing configure step for 'mimalloc_ep'
```

Appveyor passed on my fork :man_shrugging: https://ci.appveyor.com/project/wesm/arrow/builds/31856939

**Comment by wesm** [comment]
+1

**Pull request closed without merging**